beam-sync.md

```diff
@@ -0,0 +1,187 @@
+# Beam Sync
+
+## Intro
+
+Beam Sync is a new strategy to transition from an empty client node to a full-state, fully-validating
+node on the network. It borrows heavily from the current "Fast Sync" approach.
+
+#### Fast Sync
+
+This writeup assumes familiarity with Fast Sync and supporting devp2p protocols like `GetNodeData`.
+
+#### Beam Sync Benefits
+
+Beam Sync provides several important benefits, to both the syncing node, and the broader network health.
+For example:
+
+- The "Fast" syncing experience is very slow, taking days before any queries
+  or validation can be done on the new node. Beam Sync has the promise of syncing to a useable node in minutes.
+- Fast Sync nodes are pure leechers on the network. Beam Sync nodes are very effective at seeding
+  data to other Beam-syncing nodes.
+- Fast Sync gets more likely to fault as total state size grows. Beam Sync gets more likely to fault
+  as state reads per block grows.
+
+#### Beam Sync Approach
+
+Roughly, the idea is for a client to skip to the head of the chain (like Fast Sync). Unlike
+Fast Sync, the block starts executing immediately, collecting *only* the state data needed
+to execute & verify the block. Crucially, a background process collects data that is not
+being read by any block. When this background process is complete, then all of Beam Sync is
+complete, and the node transitions to Full Sync.
+
+## How to Beam Sync
+
+There are many possible approaches to Beam Sync. The plan is to evolve Beam Sync through three stages:
+
+- v0: Use existing Fast Sync protocols; works on mainnet today
+- v1: Add new devp2p protocols, to optionally accelerate Beam Sync
+- v2: Add consensus-sensitive witness proofs to block headers, to enable validation of witnesses, and enforce generation
+
+This document will focus on v0.
+
+### Overview
+
+See the path from starting with an empty node to transitioning to a Full Sync:
+
+![Flow chart](https://i.postimg.cc/3x2qtnGL/Flow-chart-v0-2.png)
+
+### Data Faults
+
+When executing the EVM and collecting data just-in-time (JIT-data), the client can detect that some data is either 1) proven to belong to the state root of the previous header, or 2) created during a local execution of the block. If neither of those is true, then a data fault is generated. The fault must describe the hash of the trie node that's missing. The fault can optionally contain the account, storage, or bytecode lookup that triggered the fault. (This will be useful in v1+ of Beam Sync)
```

Comment:
The term *fault* here is a bit ambigous.  I'm assuming you are referring to the mechanism that our current implementation uses where our state database raises an exception that allows us to see what is missing and issue a request for it over the network.  It might be nice to have a more concrete conceptual definition of this for people to wrap their heads around since this sort of bridges between spec and implementation details.

Reply:
Thoughts on this revision?

> During normal (Full Syncing) execution, a client executes the EVM using a complete state database. At each step of the executing bytecode, the client can immediately resolve any opcodes that depend on state. A `BALANCE` opcode, for example, loads an account out of the state, and the client can always verify whether the account is included in the trie, and determine what balance value to put on the stack.
>
> When Beam Syncing, a client executes the EVM using a database that is missing some data used to access state, or to prove inclusion of that state in a given state root hash. If encountering a `BALANCE` opcode, for example, there might not be enough data to prove the target account's inclusion (or exclusion) in the state trie. The same concept applies to all of: accounts, storage, and bytecode. When state is unresolvable during EVM execution, the client generates a Data Fault, and pauses EVM execution. Once the required state is downloaded, the EVM can resume and reattempt the opcode that generated the Data Fault.
>
> Implementation suggestions: The Data Fault should include the hash of the trie node that's missing. This hash can be used to request the missing trie node from peers, using `GetNodeData`. The Data Fault can optionally contain the account address or storage slot that triggered it. (This will be useful in phase 1+ of Beam Sync)